File: macos/Tests/ElevateCoreTests/ISO8601DurationTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct ISO8601DurationTests {
    @Test func parsesHoursMinutes() {
        #expect(ISO8601Duration.parse("PT8H") == .seconds(8 * 3600))
        #expect(ISO8601Duration.parse("PT30M") == .seconds(1800))
        #expect(ISO8601Duration.parse("PT1H30M") == .seconds(5400))
        #expect(ISO8601Duration.parse("P1D") == .seconds(86400))
        #expect(ISO8601Duration.parse("garbage") == nil)
    }

    @Test func formatsAsHoursAndMinutes() {
        #expect(ISO8601Duration.format(.seconds(8 * 3600)) == "PT8H")
        #expect(ISO8601Duration.format(.seconds(1800)) == "PT30M")
        #expect(ISO8601Duration.format(.seconds(5400)) == "PT1H30M")
    }

    @Test func decoderAcceptsFractionalAndPlainDates() throws {
        struct Box: Decodable { let d: Date }
        let plain = try GraphJSON.decoder.decode(Box.self, from: Data(#"{"d":"2026-09-04T08:00:00Z"}"#.utf8))
        let frac = try GraphJSON.decoder.decode(Box.self, from: Data(#"{"d":"2026-09-04T08:00:00.1234567Z"}"#.utf8))
        #expect(abs(plain.d.timeIntervalSince(frac.d)) < 1)
    }
}

File: windows/tests/Elevate.Core.Tests/Iso8601DurationTests.cs
using System.Text.Json;
using Elevate.Core.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

public class Iso8601DurationTests
{
    [Fact]
    public void ParsesHoursMinutes()
    {
        Iso8601Duration.Parse("PT8H").Should().Be(TimeSpan.FromSeconds(8 * 3600));
        Iso8601Duration.Parse("PT30M").Should().Be(TimeSpan.FromSeconds(1800));
        Iso8601Duration.Parse("PT1H30M").Should().Be(TimeSpan.FromSeconds(5400));
        Iso8601Duration.Parse("P1D").Should().Be(TimeSpan.FromSeconds(86400));
        Iso8601Duration.Parse("garbage").Should().BeNull();
    }

    [Fact]
    public void FormatsAsHoursAndMinutes()
    {
        Iso8601Duration.Format(TimeSpan.FromSeconds(8 * 3600)).Should().Be("PT8H");
        Iso8601Duration.Format(TimeSpan.FromSeconds(1800)).Should().Be("PT30M");
        Iso8601Duration.Format(TimeSpan.FromSeconds(5400)).Should().Be("PT1H30M");
    }

    [Fact]
    public void DecoderAcceptsFractionalAndPlainDates()
    {
        var plain = JsonSerializer.Deserialize<Box>("""{"d":"2026-09-04T08:00:00Z"}""", GraphJson.Options);
        var frac = JsonSerializer.Deserialize<Box>("""{"d":"2026-09-04T08:00:00.1234567Z"}""", GraphJson.Options);

        Math.Abs((plain!.D - frac!.D).TotalSeconds).Should().BeLessThan(1);
    }

    [Fact]
    public void ParseDateIsTolerantOfFractionalDigitsAndOffsets()
    {
        GraphJson.ParseDate("2026-09-04T08:00:00Z").Should().NotBeNull();
        GraphJson.ParseDate("2026-09-04T08:00:00.1Z").Should().NotBeNull();
        GraphJson.ParseDate("2026-09-04T08:00:00.1234567Z").Should().NotBeNull();
        GraphJson.ParseDate("2026-09-04T08:00:00.1234567+02:00").Should().NotBeNull();
        GraphJson.ParseDate("not a date").Should().BeNull();
    }

    private sealed record Box(DateTimeOffset D);
}
